data/tutorials/tt_02_file_manipulation.md

```diff
@@ -9,123 +9,618 @@ date: 2021-05-27T21:07:30-00:00
+
+The `open_out` and `open_out_bin` functions replace any existing file content
+with new data. Use `open_out_gen` instead if you want to preserve the existing
+content.
+
+Example:
+
+```ocaml
+let () =
+  let oc = open_out_bin "myfile.txt" in
+  output_string oc "Initial data";
+  close_out oc
+let () =
+  let oc = open_out_bin "myfile.txt" in
+  output_string oc "New data";
+  close_out oc
 ```
+
+In the example above, the second call to `open_out_bin` deletes the string
+`"Initial data"` and replaces it with `"New data"`. To preserve existing
+content, use the `open_out_gen` function with the `Open_append` flag to append
+the new content to the existing one.
+
+## Advanced Topics
+
+### File Permissions and Open Flags
+
+You can specify the behaviour of the operating system when opening a file. For
+instance, you specify the file permission of the file to be created, or whether
+to clear the content if the file already exists.
+
+Both the `In_channel` and `Out_channel` modules provide the `open_gen` function,
+which accepts a list of open flags. The
+
+Below, we'll break down each open flag and its usage:
+
+- `Open_rdonly`: opens a file in read-only mode. You can only read from the
+  file, and any attempt to write will result in an error.
+- `Open_wronly`: opens a file in write-only mode. You can write to the file, but
+  reading from the file is not allowed.
+- `Open_append`: When this flag is set, the system will always write at the end
+  of the file, appending new content instead of overwriting existing content.
+- `Open_creat`: create the file if it does not already exist.
+- `Open_trunc`: clear any existing content in the file.
+- `Open_excl`: used with `Open_creat` to ensure a new file is created. If a file
+  with the same name already exists, opening the file will fail.
+- `Open_binary`: opens the file in binary mode, which allows binary data to be
+  read from or written to the file without any conversions.
+- `Open_text`: opens the file in text mode. Depending on the system's
+  configuration, the system may perform conversions, such as replacing `'\n'`
+  with the appropriate line ending sequence for the platform.
+- `Open_nonblock`: opens the file in non-blocking mode, which allows operations
+  to return immediately instead of waiting for the operation to finish.
+
+These flags can be used in combination to provide precise control over how a
+file is opened. For example, to open a file in binary mode for reading and
+writing, allowing creation of the file if it doesn't exist, you could use
+`Open_creat` along with `Open_binary` and `Open_wronly`:
+
+```ocaml
+let oc = Out_channel.open_gen [Open_wronly; Open_creat; Open_binary] 0o666 "myfile.bin"
+```
+
+In this example, `0o666` represents the permissions set for the file (read and
+write permissions for user, group, and others), following the standard Unix
+permission format.
+
+### Reading from a File Line by Line
```

Comment:
This is not an advanced topic and it's very common. Should be moved to a higher section.